<a href="https://colab.research.google.com/github/Sarvesh1814/DeepNeuralNLP/blob/Assignment-1-Train-Word2Vec-on-peS2o-Dataset-(AllenNLP)/IT555_Deep_Neural_NLP_Assignment_1_(PyTorch_Implementation).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Importing Data

In [1]:
# Mounting Drive...
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Getting data from drive...
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/output_data.csv')
df.head()

id      source version                     added  \
0  254721023  s2ag/valid      v1  2022-12-16T16:07:05.559Z   
1  254200673  s2ag/valid      v1  2022-12-04T16:21:43.441Z   
2  252643161  s2ag/valid      v1  2022-10-01T15:18:53.788Z   
3  244852373  s2ag/valid      v1  2021-12-04T16:21:39.980Z   
4  254774363  s2ag/valid      v1  2022-12-17T16:19:10.635Z   

                    created  \
0  2022-12-14T00:00:00.000Z   
1  2022-12-02T00:00:00.000Z   
2  2022-12-01T00:00:00.000Z   
3  2022-12-01T00:00:00.000Z   
4  2022-12-14T00:00:00.000Z   

                                                text       fieldofstudy  
0  Investigation Some Characteristics of Biocompo...  Materials Science  
1  Modelling Techniques for Conjunctive Water Man...   Computer Science  
2  Complete Targets Coverage in Energy Harvesting...   Computer Science  
3  Nonorthogonal Multiple Access Enabled Two-Way ...   Computer Science  
4  106 A CASE OF COMPLETE ATRIOVENTRICULAR BLOCK ...           Medicine

### Fetching papers from Computer Science domain

In [3]:
# Now will take only those papers whose fieldofStudy == "Computer Science"

df_cs = df[df["fieldofstudy"].str.lower().str.contains("computer science")]
df_cs.head()

id      source version                     added  \
1   254200673  s2ag/valid      v1  2022-12-04T16:21:43.441Z   
2   252643161  s2ag/valid      v1  2022-10-01T15:18:53.788Z   
3   244852373  s2ag/valid      v1  2021-12-04T16:21:39.980Z   
6   252033565  s2ag/valid      v1  2022-09-03T15:09:09.625Z   
18  255290365  s2ag/valid      v1  2022-12-31T16:09:26.680Z   

                     created  \
1   2022-12-02T00:00:00.000Z   
2   2022-12-01T00:00:00.000Z   
3   2022-12-01T00:00:00.000Z   
6   2023-01-01T00:00:00.000Z   
18  2022-12-22T00:00:00.000Z   

                                                 text      fieldofstudy  
1   Modelling Techniques for Conjunctive Water Man...  Computer Science  
2   Complete Targets Coverage in Energy Harvesting...  Computer Science  
3   Nonorthogonal Multiple Access Enabled Two-Way ...  Computer Science  
6   Event-Triggered Adaptive Output Feedback Contr...  Computer Science  
18  From MMU to MPU: Adaptation of the Pip Kernel ...  Computer Science

In [4]:
# Shape of dataframe
df_cs.shape

(12421, 7)

## Preprocessing Data
1. stopwords
2. urls
3. bullets
4. apostrophe
5. hyphens
6. enumerations
7. numerical-to-text conversion
8. punctuations.

In [5]:
!pip install num2words
import re
import nltk
from nltk.corpus import stopwords
from num2words import num2words
nltk.download('stopwords')
stop_words = stopwords.words("english")

def preprocess_text(text):
    # Removing stopwords using Regex

    # stopwords = set(["a", "an", "the", "in", "on", "at", "to", "of", "for", "with", "by"])
    # text = " ".join([word for word in text.split() if word.lower() not in stopwords])

    # Removing stopwords using NLTK
    text= " ".join([word for word in text.split() if word.lower() not in stop_words])

    # Removing URLs
    text = re.sub(r'http\S+', '', text)

    # Removing bullets
    text = re.sub(r'\s•\s', ' ', text)

    # Removing apostrophes
    text = re.sub(r"'s\b", '', text)

    # Removing hyphens
    text = re.sub(r'-', ' ', text)

    # Remove enumerations
    text = re.sub(r'\b\d+\.\s*', '', text)

    # Converting numerical values to text
    text = re.sub(r'\d+', lambda x: num2words(int(x.group())), text)

    # Removing punctuation
    # text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'[^\w\s.]', '', text)
    # Converting the text to lowercase
    return text.lower()

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.2/125.2 kB 1.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13705 sha256=6885c33487795a8d32fe22978c10f4dcc4640a95469ae5464d486011a30dcbb8
  Stored in directory: /root/.cache/pip/wheels/fc/ab/d4/5da2067ac95b36618c629a5f93f809425700506f72c9732fac
Successfully built docopt


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [6]:
# Adding text_processed column to save the preprocessed data...
df_cs['text_processed'] = df_cs['text'].apply(preprocess_text)
df_cs.head()

<ipython-input-6-cca40040589d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cs['text_processed'] = df_cs['text'].apply(preprocess_text)


id      source version                     added  \
1   254200673  s2ag/valid      v1  2022-12-04T16:21:43.441Z   
2   252643161  s2ag/valid      v1  2022-10-01T15:18:53.788Z   
3   244852373  s2ag/valid      v1  2021-12-04T16:21:39.980Z   
6   252033565  s2ag/valid      v1  2022-09-03T15:09:09.625Z   
18  255290365  s2ag/valid      v1  2022-12-31T16:09:26.680Z   

                     created  \
1   2022-12-02T00:00:00.000Z   
2   2022-12-01T00:00:00.000Z   
3   2022-12-01T00:00:00.000Z   
6   2023-01-01T00:00:00.000Z   
18  2022-12-22T00:00:00.000Z   

                                                 text      fieldofstudy  \
1   Modelling Techniques for Conjunctive Water Man...  Computer Science   
2   Complete Targets Coverage in Energy Harvesting...  Computer Science   
3   Nonorthogonal Multiple Access Enabled Two-Way ...  Computer Science   
6   Event-Triggered Adaptive Output Feedback Contr...  Computer Science   
18  From MMU to MPU: Adaptation of the Pip Kernel ...  Computer Science   

                                       text_processed  
1   modelling techniques conjunctive water managem...  
2   complete targets coverage energy harvesting in...  
3   nonorthogonal multiple access enabled two way ...  
6   event triggered adaptive output feedback contr...  
18  mmu mpu adaptation pip kernel constrained devi...

## Random training data of 3000 documents are used, while a test dataset of 1000 documents is to be used. (Train Test Splitting)

**We will consider first 4000 rows only for our task and then we will split it into train and test set. Currently we have 12421 rows in the dataset**

In [7]:
data = df_cs.iloc[:4000]

In [8]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(data, test_size=1000, random_state=42)

print("Train data shape:", train_data.shape)
print("Test data shape:", test_data.shape)

Train data shape: (3000, 8)
Test data shape: (1000, 8)


## **Getting Data**

In [9]:
%%capture
!pip install torch torchtext torchdata

In [10]:
import os
import random
import re
from collections import Counter, OrderedDict
from dataclasses import dataclass
from time import monotonic
from typing import Dict, List, Optional, Union

import numpy as np
import torch
import torch.nn as nn
from scipy.spatial.distance import cosine
from torch.utils.data import DataLoader
from torchtext.data import to_map_style_dataset
from torchtext.data.utils import get_tokenizer
from torchtext.datasets import WikiText103
from tqdm import tqdm

In [11]:
from torchtext.data import to_map_style_dataset

In [12]:
def get_data():
    # gets the data
    train_iter = to_map_style_dataset(train_data)
    test_iter = to_map_style_dataset(test_data)
    return train_iter, test_iter

train_iter, test_iter = get_data()

## **Setting Parameters and Configuration**

In [13]:
from dataclasses import dataclass, field
import torch
import torch.nn as nn

In [28]:
@dataclass(repr=True)
class Word2VecParams:

    # skipgram parameters
    MIN_FREQ = 1
    SKIPGRAM_N_WORDS = 8
    T = 85
    NEG_SAMPLES = 50
    NS_ARRAY_LEN = 5_000_000
    SPECIALS = ""
    TOKENIZER = 'basic_english'

    # network parameters
    BATCH_SIZE = 100
    EMBED_DIM = 300
    EMBED_MAX_NORM = None
    N_EPOCHS = 5
    DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    CRITERION = nn.BCELoss()

In [15]:
params = Word2VecParams()

## **Building the Vocabulary**

In [16]:
class Vocab:
    def __init__(self, list, specials):
        self.stoi = {v[0]:(k, v[1]) for k, v in enumerate(list)}
        self.itos = {k:(v[0], v[1]) for k, v in enumerate(list)}
        self._specials = specials[0]
        self.total_tokens = np.nansum(
            [f for _, (_, f) in self.stoi.items()]
            , dtype=int)

    def __len__(self):
        return len(self.stoi) - 1

    def get_index(self, word: Union[str, List]):
        if isinstance(word, str):
            if word in self.stoi:
                return self.stoi.get(word)[0]
            else:
                return self.stoi.get(self._specials)[0]
        elif isinstance(word, list):
            res = []
            for w in word:
                if w in self.stoi:
                    res.append(self.stoi.get(w)[0])
                else:
                    res.append(self.stoi.get(self._specials)[0])
            return res
        else:
            raise ValueError(
                f"Word {word} is not a string or a list of strings."
                )


    def get_freq(self, word: Union[str, List]):
        if isinstance(word, str):
            if word in self.stoi:
                return self.stoi.get(word)[1]
            else:
                return self.stoi.get(self._specials)[1]
        elif isinstance(word, list):
            res = []
            for w in word:
                if w in self.stoi:
                    res.append(self.stoi.get(w)[1])
                else:
                    res.append(self.stoi.get(self._specials)[1])
            return res
        else:
            raise ValueError(
                f"Word {word} is not a string or a list of strings."
                )


    def lookup_token(self, token: Union[int, List]):
        if isinstance(token, (int, np.int64)):
            if token in self.itos:
                return self.itos.get(token)[0]
            else:
                raise ValueError(f"Token {token} not in vocabulary")
        elif isinstance(token, list):
            res = []
            for t in token:
                if t in self.itos:
                    res.append(self.itos.get(token)[0])
                else:
                    raise ValueError(f"Token {t} is not a valid index.")
            return res

In [19]:
def yield_tokens(iterator, tokenizer):
    r = re.compile('[a-z1-9]')
    for text in iterator:
        res = tokenizer(text)
        res = list(filter(r.match, res))
        yield res

def vocab(ordered_dict: Dict, min_freq: int = 1, specials: str = ''):
    tokens = []
    # Save room for special tokens
    for token, freq in ordered_dict.items():
        if freq >= min_freq:
            tokens.append((token, freq))

    specials = (specials, np.nan)
    tokens[0] = specials

    return Vocab(tokens, specials)

def pipeline(word, vocab, tokenizer):
    return vocab(tokenizer(word))

def build_vocab(
        iterator,
        tokenizer,
        params: Word2VecParams,
        max_tokens: Optional[int] = None,
    ):
    counter = Counter()
    for tokens in yield_tokens(iterator, tokenizer):
        counter.update(tokens)

    # First sort by descending frequency, then lexicographically
    sorted_by_freq_tuples = sorted(
        counter.items(), key=lambda x: (-x[1], x[0])
        )

    ordered_dict = OrderedDict(sorted_by_freq_tuples)

    word_vocab = vocab(
        ordered_dict, min_freq=params.MIN_FREQ, specials=params.SPECIALS
        )
    return word_vocab

In [20]:
class SkipGrams:
    def __init__(self, vocab: Vocab, params: Word2VecParams, tokenizer):
        self.vocab = vocab
        self.params = params
        self.t = self._t()
        self.tokenizer = tokenizer
        self.discard_probs = self._create_discard_dict()

    def _t(self):
        freq_list = []
        for _, (_, freq) in list(self.vocab.stoi.items())[1:]:
            freq_list.append(freq/self.vocab.total_tokens)
        return np.percentile(freq_list, self.params.T)


    def _create_discard_dict(self):
        discard_dict = {}
        for _, (word, freq) in self.vocab.stoi.items():
            dicard_prob = 1-np.sqrt(
                self.t / (freq/self.vocab.total_tokens + self.t))
            discard_dict[word] = dicard_prob
        return discard_dict


    def collate_skipgram(self, batch):
        batch_input, batch_output  = [], []
        for text in batch:
            text_tokens = self.vocab.get_index(self.tokenizer(text))

            if len(text_tokens) < self.params.SKIPGRAM_N_WORDS * 2 + 1:
                continue

            for idx in range(len(text_tokens) - self.params.SKIPGRAM_N_WORDS*2
                ):
                token_id_sequence = text_tokens[
                    idx : (idx + self.params.SKIPGRAM_N_WORDS * 2 + 1)
                    ]
                input_ = token_id_sequence.pop(self.params.SKIPGRAM_N_WORDS)
                outputs = token_id_sequence

                prb = random.random()
                del_pair = self.discard_probs.get(input_)
                if input_==0 or del_pair >= prb:
                    continue
                else:
                    for output in outputs:
                        prb = random.random()
                        del_pair = self.discard_probs.get(output)
                        if output==0 or del_pair >= prb:
                            continue
                        else:
                            batch_input.append(input_)
                            batch_output.append(output)

        batch_input = torch.tensor(batch_input, dtype=torch.long)
        batch_output = torch.tensor(batch_output, dtype=torch.long)

        return batch_input, batch_output

In [21]:
class NegativeSampler:
    def __init__(self, vocab: Vocab, ns_exponent: float, ns_array_len: int):
        self.vocab = vocab
        self.ns_exponent = ns_exponent
        self.ns_array_len = ns_array_len
        self.ns_array = self._create_negative_sampling()

    def __len__(self):
        return len(self.ns_array)

    def _create_negative_sampling(self):

        frequency_dict = {word:freq**(self.ns_exponent) \
                          for _,(word, freq) in
                          list(self.vocab.stoi.items())[1:]}
        frequency_dict_scaled = {
            word:
            max(1,int((freq/self.vocab.total_tokens)*self.ns_array_len))
            for word, freq in frequency_dict.items()
            }
        ns_array = []
        for word, freq in tqdm(frequency_dict_scaled.items()):
            ns_array = ns_array + [word]*freq
        return ns_array

    def sample(self,n_batches: int=1, n_samples: int=1):
        samples = []
        for _ in range(n_batches):
            samples.append(random.sample(self.ns_array, n_samples))
        samples = torch.as_tensor(np.array(samples))
        return samples

In [22]:
class Model(nn.Module):
    def __init__(self, vocab: Vocab, params: Word2VecParams):
        super().__init__()
        self.vocab = vocab
        self.t_embeddings = nn.Embedding(
            self.vocab.__len__()+1,
            params.EMBED_DIM,
            max_norm=params.EMBED_MAX_NORM
            )
        self.c_embeddings = nn.Embedding(
            self.vocab.__len__()+1,
            params.EMBED_DIM,
            max_norm=params.EMBED_MAX_NORM
            )

    def forward(self, inputs, context):
        # getting embeddings for target & reshaping
        target_embeddings = self.t_embeddings(inputs)
        n_examples = target_embeddings.shape[0]
        n_dimensions = target_embeddings.shape[1]
        target_embeddings = target_embeddings.view(n_examples, 1, n_dimensions)

        # get embeddings for context labels & reshaping
        # Allows us to do a bunch of matrix multiplications
        context_embeddings = self.c_embeddings(context)
        # * This transposes each batch
        context_embeddings = context_embeddings.permute(0,2,1)

        # * custom linear layer
        dots = target_embeddings.bmm(context_embeddings)
        dots = dots.view(dots.shape[0], dots.shape[2])
        return dots

    def normalize_embeddings(self):
        embeddings = list(self.t_embeddings.parameters())[0]
        embeddings = embeddings.cpu().detach().numpy()
        norms = (embeddings ** 2).sum(axis=1) ** (1 / 2)
        norms = norms.reshape(norms.shape[0], 1)
        return embeddings / norms

    def get_similar_words(self, word, n):
        word_id = self.vocab.get_index(word)
        if word_id == 0:
            print("Out of vocabulary word")
            return

        embedding_norms = self.normalize_embeddings()
        word_vec = embedding_norms[word_id]
        word_vec = np.reshape(word_vec, (word_vec.shape[0], 1))
        dists = np.matmul(embedding_norms, word_vec).flatten()
        topN_ids = np.argsort(-dists)[1 : n + 1]

        topN_dict = {}
        for sim_word_id in topN_ids:
            sim_word = self.vocab.lookup_token(sim_word_id)
            topN_dict[sim_word] = dists[sim_word_id]
        return topN_dict

    def get_similarity(self, word1, word2):
        idx1 = self.vocab.get_index(word1)
        idx2 = self.vocab.get_index(word2)
        if idx1 == 0 or idx2 == 0:
            print("One or both words are out of vocabulary")
            return

        embedding_norms = self.normalize_embeddings()
        word1_vec, word2_vec = embedding_norms[idx1], embedding_norms[idx2]

        return cosine(word1_vec, word2_vec)

In [23]:

class Trainer:
    def __init__(self, model: Model, params: Word2VecParams, optimizer,
                vocab: Vocab, train_iter, valid_iter, skipgrams: SkipGrams):
        self.model = model
        self.optimizer = optimizer
        self.vocab = vocab
        self.train_iter = train_iter
        self.valid_iter = valid_iter
        self.skipgrams = skipgrams
        self.params = params

        self.epoch_train_mins = {}
        self.loss = {"train": [], "valid": []}

        # sending all to device
        self.model.to(self.params.DEVICE)
        self.params.CRITERION.to(self.params.DEVICE)

        self.negative_sampler = NegativeSampler(
            vocab=self.vocab, ns_exponent=.75,
            ns_array_len=self.params.NS_ARRAY_LEN
            )
        self.testwords = ['love', 'hurricane', 'military', 'army']


    def train(self):
        self.test_testwords()
        for epoch in range(self.params.N_EPOCHS):
            # Generate Dataloaders
            self.train_dataloader = DataLoader(
                self.train_iter,
                batch_size=self.params.BATCH_SIZE,
                shuffle=False,
                collate_fn=self.skipgrams.collate_skipgram
            )
            self.valid_dataloader = DataLoader(
                self.valid_iter,
                batch_size=self.params.BATCH_SIZE,
                shuffle=False,
                collate_fn=self.skipgrams.collate_skipgram
            )
            # training the model
            st_time = monotonic()
            self._train_epoch()
            self.epoch_train_mins[epoch] = round((monotonic()-st_time)/60, 1)

            # validating the model
            self._validate_epoch()
            print(f"""Epoch: {epoch+1}/{self.params.N_EPOCHS}\n""",
            f"""    Train Loss: {self.loss['train'][-1]:.2}\n""",
            f"""    Valid Loss: {self.loss['valid'][-1]:.2}\n""",
            f"""    Training Time (mins): {self.epoch_train_mins.get(epoch)}"""
            """\n"""
            )
            self.test_testwords()


    def _train_epoch(self):
        self.model.train()
        running_loss = []

        for i, batch_data in enumerate(self.train_dataloader, 1):
            if len(batch_data[0]) == 0:
                continue
            inputs = batch_data[0].to(self.params.DEVICE)
            pos_labels = batch_data[1].to(self.params.DEVICE)
            neg_labels = self.negative_sampler.sample(
                pos_labels.shape[0], self.params.NEG_SAMPLES
                )
            neg_labels = neg_labels.to(self.params.DEVICE)
            context = torch.cat(
                [pos_labels.view(pos_labels.shape[0], 1),
                neg_labels], dim=1
              )

            # building the targets tensor
            y_pos = torch.ones((pos_labels.shape[0], 1))
            y_neg = torch.zeros((neg_labels.shape[0], neg_labels.shape[1]))
            y = torch.cat([y_pos, y_neg], dim=1).to(self.params.DEVICE)

            self.optimizer.zero_grad()

            outputs = self.model(inputs, context)
            loss = self.params.CRITERION(outputs, y)
            loss.backward()
            self.optimizer.step()

            running_loss.append(loss.item())

        epoch_loss = np.mean(running_loss)

        self.loss['train'].append(epoch_loss)

    def _validate_epoch(self):
        self.model.eval()
        running_loss = []

        with torch.no_grad():
            for i, batch_data in enumerate(self.valid_dataloader, 1):
                if len(batch_data[0]) == 0:
                    continue
                inputs = batch_data[0].to(self.params.DEVICE)
                pos_labels = batch_data[1].to(self.params.DEVICE)
                neg_labels = self.negative_sampler.sample(
                    pos_labels.shape[0], self.params.NEG_SAMPLES
                    ).to(self.params.DEVICE)
                context = torch.cat(
                    [pos_labels.view(pos_labels.shape[0], 1),
                    neg_labels], dim=1
                  )


                # building the targets tensor
                y_pos = torch.ones((pos_labels.shape[0], 1))
                y_neg = torch.zeros((neg_labels.shape[0], neg_labels.shape[1]))
                y = torch.cat([y_pos, y_neg], dim=1).to(self.params.DEVICE)

                preds = self.model(inputs, context).to(self.params.DEVICE)
                loss = self.params.CRITERION(preds, y)

                running_loss.append(loss.item())

            epoch_loss = np.mean(running_loss)
            self.loss['valid'].append(epoch_loss)

    def test_testwords(self, n: int = 5):
        for word in self.testwords:
            print(word)
            nn_words = self.model.get_similar_words(word, n)
            for w, sim in nn_words.items():
                print(f"{w} ({sim:.3})", end=' ')
            print('\n')

In [26]:
params = Word2VecParams()
train_iter, valid_iter = get_data()
tokenizer = get_tokenizer(params.TOKENIZER)
vocab = build_vocab(train_iter, tokenizer, params)
skip_gram = SkipGrams(vocab=vocab, params=params, tokenizer=tokenizer)
model = Model(vocab=vocab, params=params).to(params.DEVICE)
optimizer = torch.optim.Adam(params = model.parameters())

In [29]:
trainer = Trainer(
        model=model,
        params=params,
        optimizer=optimizer,
        train_iter=train_iter,
        valid_iter=valid_iter,
        vocab=vocab,
        skipgrams=skip_gram
    )
trainer.train()

100%|██████████| 7/7 [00:00<00:00, 22.26it/s]


love
Out of vocabulary word


AttributeError: ignored